### DISCUSSION TITLE
Is there a way to first recieve with tag and then choose the type?

### DISCUSSION THREAD

**Discussion by lquenti** [rsmpi/rsmpi#165] [Q&A] [answered]
Hey! I am building sth with a single coordinator/multiple worker architecture and have would love to do something along the lines of

```
const FOO_DATASTRUCTURE_TAG: Rank = 0;
const BAR_DATASTRUCTURE_TAG: Rank = 1;

...

if rank == ROOT {
  let (msg_of_unknown_type, status) = world.recieve_any_with_tag();
  let msg = match status.tag {
    FOO_DATASTRUCTURE_TAG => msg_of_unknown_type.cast_into<DataStructureFoo>(),
    BAR_DATASTRUCTURE_TAG => msg_of_unknown_type.cast_into<DataStructureBar>(),
  }
} else {
  // so some logic
  if some_condition {
    world.process_by_rank(ROOT).send_with_tag(&foo, FOO_DATASTRUCTURE_TAG);
  } else {
    world.process_by_rank(ROOT).send_with_tag(&bar, BAR_DATASTRUCTURE_TAG);
  }
}
```

How do I archive this? After thoroughly looking through the docs, I have not found anything.

I now opt out with sending a "tag only" message first, on which I then match on the root process. But this is very inefficient as it doubles the RTT time!

Thank you for your answer, I am very grateful and to provide any further information.

**Comment by hppritcha** [comment, marked as answer]
Have you looked into using the probe method for checking received messages?

**Reply by lquenti** [reply]
I am so sorry and embarrassed, I do not know how I could have missed that.

Thank you a lot.
